runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -278,24 +347,73 @@ public String backendWorkerToken() {
   }
 
   @Override
-  public void shutdown() {
+  public final void shutdown() {
```

Comment:
we do not want to synchronize on this since there are other stream operations that will block shutdown in that case (which we do not want)

we want it to act like a signal

maybe we can synchronize on isShutdown for start and shutdown methods

Reply:
added a shutdownLock to only be used here